src/compute-types/src/plan/flat_plan.rs

```diff
@@ -0,0 +1,1148 @@
+    Union {
+        /// The input collections.
+        inputs: Vec<NodeId>,
+        /// Whether to consolidate the output, e.g., cancel negated records.
+        consolidate_output: bool,
+    },
+    /// The `input` plan, but with additional arrangements.
+    ///
+    /// This operator does not change the logical contents of `input`, but ensures that certain
+    /// arrangements are available in the results. This operator can be important for e.g. the
+    /// `Join` stage which benefits from multiple arrangements or to cap a `Plan` so that indexes
+    /// can be exported.
+    ArrangeBy {
+        /// The input collection.
+        input: NodeId,
+        /// A list of arrangement keys, and possibly a raw collection, that will be added to those
+        /// of the input.
+        ///
+        /// If any of these collection forms are already present in the input, they have no effect.
+        forms: AvailableCollections,
+        /// The key that must be used to access the input.
+        input_key: Option<Vec<MirScalarExpr>>,
+        /// The MFP that must be applied to the input.
+        input_mfp: MapFilterProject,
+    },
+}
+
+impl<T> From<Plan<T>> for FlatPlan<T> {
+    /// Flatten the given [`Plan`] into a [`FlatPlan`].
+    fn from(plan: Plan<T>) -> Self {
+        // The strategy is to walk walk through the `Plan` in pre-order and at each step (a)
+        // produce a corresponding `FlatPlanNode` and (b) push next steps for the contained
+        // subplans onto the work stack. We do this until no further steps remain.
+        //
+        // Special cases exist for `Let`s and `Get`s: We want to remove the former and let the
+        // latter reference values by their `NodeId`s directly. For example:
+        //
+        //         +-------+                             +-------+
+        //         | Root  |                             | Root  |
+        //         +-------+                             +-------+
+        //             |                                     |
+        //         +-------+                             +-------+
+        //         |  Let  |                             | Body  |
+        //         +-------+                             +-------+
+        //           /   \                 ===>            /   \
+        //       /---     ---\                          /--     --\
+        //  +-------+     +-------+                +-------+   +-------+
+        //  | Value |     | Body  |                |  Get  |   |  Get  |
+        //  +-------+     +-------+                +-------+   +-------+
+        //                  /   \                       \         /
+        //               /--     --\                     --\   /--
+        //          +-------+   +-------+                +-------+
+        //          |  Get  |   |  Get  |                | Value |
+        //          +-------+   +-------+                +-------+
```

Comment:
It might be possible to remove the `Get` nodes as well. But `Get`s can encode logic beyond just looking up a collection (see `GetPlan`) and we'd need to assert that this isn't the case for local `Get`s, which I'm not sure we can.